terraform_providers_lock.sh

```diff
@@ -47,42 +72,74 @@ parse_cmdline_() {
   done
 }
 
-terraform_providers_lock_() {
-  local -a paths
-  local index=0
-  local file_with_path
+function common::per_dir_hook {
+  local -r args="$1"
+  shift 1
+  local -a -r files=("$@")
 
-  for file_with_path in "${FILES[@]}"; do
+  # consume modified files passed from pre-commit so that
+  # hook runs against only those relevant directories
+  local index=0
+  for file_with_path in "${files[@]}"; do
     file_with_path="${file_with_path// /__REPLACED__SPACE__}"
 
-    paths[index]=$(dirname "$file_with_path")
+    dir_paths[index]=$(dirname "$file_with_path")
 
     ((index += 1))
   done
 
-  local path_uniq
-  for path_uniq in $(echo "${paths[*]}" | tr ' ' '\n' | sort -u); do
-    path_uniq="${path_uniq//__REPLACED__SPACE__/ }"
-
-    if [[ ! -d "${path_uniq}/.terraform" ]]; then
-      set +e
-      init_output=$(terraform -chdir="${path_uniq}" init -backend=false 2>&1)
-      init_code=$?
-      set -e
-
-      if [[ $init_code != 0 ]]; then
-        echo "Init before validation failed: $path_uniq"
-        echo "$init_output"
-        exit 1
-      fi
+  # allow hook to continue if exit_code is greater than 0
+  # preserve errexit status
+  shopt -qo errexit && ERREXIT_IS_SET=true
+  set +e
+  local final_exit_code=0
+
+  # run hook for each path
+  for dir_path in $(echo "${dir_paths[*]}" | tr ' ' '\n' | sort -u); do
+    dir_path="${dir_path//__REPLACED__SPACE__/ }"
+    pushd "$dir_path" > /dev/null
+
+    per_dir_hook_unique_part "$args" "$dir_path"
+
+    local exit_code=$?
+    if [ "$exit_code" != 0 ]; then
+      final_exit_code=$exit_code
     fi
 
-    terraform -chdir="${path_uniq}" providers lock "${ARGS[@]}"
+    popd > /dev/null
   done
+
+  # restore errexit if it was set before the "for" loop
+  [[ $ERREXIT_IS_SET ]] && set -e
+  # return the hook final exit_code
+  exit $final_exit_code
 }
 
-# global arrays
-declare -a ARGS
-declare -a FILES
+function per_dir_hook_unique_part {
+  # common logic located in common::per_dir_hook
+  local -r args="$1"
+  local -r dir_path="$2"
+
+  if [[ ! -d ".terraform" ]]; then
+    set +e
+    init_output=$(terraform init -backend=false 2>&1)
+    init_code=$?
+    set -e
+
+    if [[ $init_code != 0 ]]; then
+      common::colorify "red" "Init before validation failed: $dir_path"
+      common::colorify "red" "$init_output"
+      exit 1
```

Comment:
Might it worth to return exit code of `terraform init` rather then a somewhat synthetic exit code?
```suggestion
      exit $init_code
```